include/rdma/nccl_ofi_rdma_request.h

```diff
@@ -338,21 +338,128 @@ typedef struct nccl_net_ofi_rdma_req {
 
 	/*
 	 * @brief	Return eager local copy data struct of request
-	 */	
+	 */
 	rdma_req_eager_copy_data_t *get_eager_copy_data();
-	
+
 
 	/*
 	 * @brief	Return receive segments data struct of receive segments request
 	 */
 	rdma_req_recv_segms_data_t *get_recv_segms_data();
-	
+
 
 	/*
- 	 * @brief	Return flush data struct of flush request
- 	 */
+	 * @brief	Return flush data struct of flush request
+	 */
 	rdma_req_flush_data_t *get_flush_data();
 
+
+	/*
+	 * @brief 	Increment request completions of main requests and set request
+	 *		state to completed if total number of completions is reached
+	 *
+	 * Note that the request state is only updated if the request state
+	 * does not track an error already.
+	 *
+	 * This function is guarded by the request lock.
+	 *
+	 * To update the state of subrequests, use the subrequest specific
+	 * update functions.
+	 *
+	 * @param	size
+	 *		Size of the completion
+	 * @param	total_ncompls
+	 *		Total number of expected request completions
+	 * @return	0, on success
+	 *		non-zero, on error
+	 */
+	int inc_req_completion(size_t size, int total_ncompls);
+
+
+	/*
+	 * @brief	Set ctrl request to completed
+	 *
+	 * Set send ctrl request to completed. Furthermore, increment
+	 * completions of parent request (receive request).
+	 *
+	 * Modifications of the send control request are guarded by the send
+	 * control request's lock.  Modifications of the receive request are
+	 * guarded by the receive request's lock.
+	 *
+	 * @return	0, on success
+	 *		non-zero, on error
+	 */
+	int set_send_ctrl_completed();
+
+
+	/*
+	 * @brief	Increment segment completions of receive segment request
+	 *
+	 * Increment segment completions of receive segment request. In case
+	 * all segments arrived, increment completions of parent request
+	 * (receive request).
+	 *
+	 * Modifications of the receive segment request are guarded by the
+	 * receive segment request's lock.  Modifications of the receive
+	 * request are guarded by the receive request's lock.
+	 *
+	 * @param	size
+	 *		Size of the completed segment
+	 * @param	total_nsegms
+	 *		Total number of expected segments
+	 * @return	0, on success
+	 *		non-zero, on error
+	 */
+	int inc_recv_seg_completion(size_t size, int total_nsegms);
+
+
+	/**
+	 * @brief	Handle completion for a flush request
+	 */
+	int handle_flush_comp();
+
+
+
+	/*
+	 * @brief	Zero out rdma request
+	 */
+	void zero_nccl_ofi_req();
+
+
+	/**
+	 * @brief	Set state of request and potential parent requests to error
+	 */
+	void set_request_state_to_error();
+
+
+	/**
+	 * @brief	Free request by returning request back into freelist
+	 */
+	int free_base_req(uint64_t *num_inflight_reqs,
```

Comment:
this should be a destructor?

Reply:
A problem with making this a destructor is that we can't pass in arguments to a destructor, and thus can't pass in the freelist as an argument to make the `nccl_ofi_freelist_entry_free` or pass in the `num_inflight_reqs` arg to decrement. I assume we will need to either change `nccl_net_ofi_rdma_req_t` to have internal access to the freelist and `num_inflight_reqs` pointers, or move the freelist and `num_inflight_reqs` updates outside the destructor.